tests/atomistics/job/test_structurecontainer.py

```diff
@@ -0,0 +1,110 @@
+        self.assertEqual(sorted(self.elements), sorted(self.cont.get_elements()),
+                         "Results from get_elements() do not match added elements.")
+
+    def test_add_array(self):
+        """Custom arrays added with add_array should be properly allocated with matching shape, dtype and fill"""
+
+        self.cont.add_array("energy", per="structure")
+        self.cont.add_array("forces", shape=(3,), per="atom")
+        self.cont.add_array("fnorble", shape=(), dtype=np.int64, fill=0, per="atom")
+
+        self.assertTrue("energy" in self.cont._per_structure_arrays,
+                        "no 'energy' array present after adding it with add_array()")
+        self.assertEqual(self.cont._per_structure_arrays["energy"].shape, (self.cont._num_structures_alloc,),
+                        "'energy' array has wrong shape")
+
+        self.assertTrue("forces" in self.cont._per_atom_arrays,
+                        "no 'forces' array present after adding it with add_array()")
+        self.assertEqual(self.cont._per_atom_arrays["forces"].shape, (self.cont._num_atoms_alloc, 3),
+                        "'forces' array has wrong shape")
+
+        self.assertEqual(self.cont._per_atom_arrays["fnorble"].dtype, np.int64,
+                         "'fnorble' array has wrong dtype after adding it with add_array()")
+        self.assertTrue((self.cont._per_atom_arrays["fnorble"] == 0).all(),
+                         "'fnorble' array not initialized with given fill value.")
+
+    def test_get_structure(self):
+        """Structure from get_structure should match thoes add with add_structure exactly."""
+
+        for i, s in enumerate(self.structures):
+            self.assertEqual(s, self.cont.get_structure(i),
+                             "Added structure not equal to returned structure.")
+
+    def test_add_structure_kwargs(self):
+        """Additional kwargs given to add_structure should create appropriate custom arrays."""
+
+        E = 3.14
+        P = np.eye(3) * 2.72
+        F = np.array([[1,3,5]] * len(self.structures[0]))
+        R = np.ones(len(self.structures[0]))
+        self.cont.add_structure(self.structures[0], self.structures[0].get_chemical_formula(),
+                                energy=E, forces=F, pressure=P, fnord=R[None, :])
+        self.assertEqual(self.cont.get_array("energy", self.cont.num_structures - 1), E,
+                         "Energy returned from get_array() does not match energy passed to add_structure")
+        self.assertTrue(np.allclose(self.cont.get_array("forces", self.cont.num_structures - 1), F),
+                        "Forces returned from get_array() does not match forces passed to add_structure")
+        self.assertTrue(np.allclose(self.cont.get_array("pressure", self.cont.num_structures - 1), P),
+                        "Pressure returned from get_array() does not match pressure passed to add_structure")
+        self.assertTrue("fnord" in self.cont._per_structure_arrays,
```

Comment:
for QoL I think it's nice to `f'print the {thing} that was wrong'` in these more general assertions that wouldn't otherwise print it, e.g. `self.cont._per_structure_arrays.keys()` or w/e here.